File: go/languagecheck/translate.go
/*
	Copyright 2015, Google, Inc.
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/
// This file contains all of the logic to talk to Google Translate API and handle caching results.
package languagecheck

import (
	"encoding/json"
	"errors"
	"io/ioutil"
	"net/url"

	"appengine"
	"appengine/memcache"
	"appengine/urlfetch"
)

const (
	defaultTarget = "en"
	base          = "https://www.googleapis.com/language/translate/v2"
)

// The format for the list of all languages in a target language.
type langList struct {
	Data struct {
		Languages []struct {
			Language string
			Name     string
		}
	}
}

// Format for getting detection information out of Google Translate
type dList struct {
	Data struct {
		Detections [][]struct {
			Language string
		}
	}
}

// The code + target language version of the language we want to return to service
type Result struct {
	Data  string `json:"data"`
	Label string `json:"label"`
}

// Returns the human friendly label for a language, in the target language
func Label(lang, target string, c appengine.Context) (string, error) {
	key := "lang:" + target + ":" + lang

	item, err := memcache.Get(c, key)
	if err != nil && err != memcache.ErrCacheMiss {
		return "", err
	}

	var result string
	if err == memcache.ErrCacheMiss {

		lList, err := langFromAPI(target, c)
		if err != nil {
			return "", err
		}

		for key, value := range lList {
			k := "lang:" + target + ":" + key

			item := &memcache.Item{
				Key:   k,
				Value: []byte(value),
			}

			memcache.Set(c, item)
		}

		r, ok := lList[lang]
		if !ok {
			return "", errors.New("Language not supported")
		}
		result = string(r)

	} else {
		value := item.Value
		result = string(value)
	}

	return result, nil
}

// Returns the human friendly label and language code for the input string: s
func Detect(s, target string, c appengine.Context) (Result, error) {

	item, err := memcache.Get(c, s)
	if err != nil && err != memcache.ErrCacheMiss {
		return Result{"", ""}, err
	}

	var r string
	if err == memcache.ErrCacheMiss {
		r, err = detectFromAPI(s, c)
		if err != nil {
			return Result{"", ""}, err
		}
		item := &memcache.Item{
			Key:   s,
			Value: []byte(r),
		}

		memcache.Set(c, item)

	} else {
		value := item.Value
		r = string(value)
	}

	l, err := Label(r, target, c)
	if err != nil {
		return Result{"", ""}, err
	}

	result := Result{r, l}

	return result, nil
}

// Makes a call to Google Translate API to get the language of the input string
func detectFromAPI(s string, c appengine.Context) (string, error) {

	path := "/detect"

	var u *url.URL
	u, err := url.Parse(base + path)
	if err != nil {
		return "", err
	}

	parameters := url.Values{}
	parameters.Add("q", s)
	parameters.Add("key", cfg.APP.Key)
	u.RawQuery = parameters.Encode()

	client := urlfetch.Client(c)

	resp, err := client.Get(u.String())
	if err != nil {
		return "", err
	}
	body, err := ioutil.ReadAll(resp.Body)
	resp.Body.Close()
	if err != nil {
		return "", err
	}

	var d dList
	if err := json.Unmarshal(body, &d); err != nil {
		return "", err
	}

	return d.Data.Detections[0][0].Language, nil
}

// Makes a call to Google Translate API to get the list of human readable languages in the target language
func langFromAPI(target string, c appengine.Context) (map[string]string, error) {
	path := "/languages"
	q := "?target=" + target + "&key=" + cfg.APP.Key
	url := base + path + q
	client := urlfetch.Client(c)

	resp, err := client.Get(url)
	if err != nil {
		return nil, err
	}
	body, err := ioutil.ReadAll(resp.Body)
	resp.Body.Close()
	if err != nil {
		return nil, err
	}

	var l langList
	if err := json.Unmarshal(body, &l); err != nil {
		return nil, err
	}

	result := make(map[string]string)
	for _, value := range l.Data.Languages {
		result[value.Language] = value.Name
	}

	return result, nil
}

File: java/LanguageCheck/src/com/google/cloud/das/languagecheck/Translate.java
/*
	Copyright 2015, Google, Inc.
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/
/**
 * 
 */
package com.google.cloud.das.languagecheck;


import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.util.Enumeration;
import java.util.Hashtable;

import com.google.appengine.api.memcache.MemcacheService;
import com.google.appengine.api.memcache.MemcacheServiceFactory;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;

/**
 * @author tpryan
 *
 */
public class Translate {
	private String key;
	private String target;
	private static final String baseurl = "https://www.googleapis.com/language/translate/v2";
	public static final String default_target = "en";
	
	public Translate(String key, String target) {
		this.key = key;
		this.target = target;
	}	
	
	public Translate(String key) {
		this(key, default_target);
	}	
	
	public Language detect(String str){
		String key = str;
		String code = "";
		
		// Using the synchronous cache
	    MemcacheService syncCache = MemcacheServiceFactory.getMemcacheService();
	    //syncCache.setErrorHandler(ErrorHandlers.getConsistentLogAndContinue(Level.INFO));
	    code = (String) syncCache.get(key); // read from cache
	    if (code == null) {
	    	System.out.println("Cache miss");
	    	code = detectFromAPI(key);
	    	syncCache.put(key, code);
	    } else{
	    	System.out.println("Cache hit");
	    }
			
	    String label = getLangaugeLabel(code);
	    Language lang = new Language(code, label);
	    
		return lang;
		
	}
	
	public String detectFromAPI(String str){
		String path = "/detect";
		String query = "";
		try {
			query = "?q=" + URLEncoder.encode(str,"UTF-8") + "&key=" + key;
		} catch (UnsupportedEncodingException e1) {
			// TODO Auto-generated catch block
			e1.printStackTrace();
		}
		String url = baseurl + path + query;
		String response = "";
		try {
			response = sendGet(url);
			//System.out.println(response);
        } catch (Exception e) {
            //do something clever with the exception
        	response = "{}";
            System.out.println(e.getMessage());
        }
		
		
		JsonParser jsonParser = new JsonParser();
		String result = jsonParser.parse(response)
						.getAsJsonObject()
						.getAsJsonObject("data")
						.getAsJsonArray("detections").get(0)
						.getAsJsonArray().get(0)
						.getAsJsonObject().get("language")
						.getAsString();
						
		
		return result;
	}
	
	public String getLangaugeLabel(String lang){
		String key = "lang:" +  this.target + ":" +  lang;
		String result = "";
		
		
		// Using the synchronous cache
	    MemcacheService syncCache = MemcacheServiceFactory.getMemcacheService();
	    //syncCache.setErrorHandler(ErrorHandlers.getConsistentLogAndContinue(Level.INFO));
	    result = (String) syncCache.get(key); // read from cache
	    if (result == null) {
	    	System.out.println("Cache miss");
	    	Hashtable<String, String> hash = getLanguagesFromAPI();
	    	result = hash.get(lang);
	    	// populate cache
	    	Enumeration<String> en = hash.keys();
	    	while(en.hasMoreElements()){
	    		String code = en.nextElement();
	    		String tempkey = "lang:" +  this.target + ":" +  code;
	    		syncCache.put(tempkey, hash.get(code));
	    	}
	    } else{
	    	System.out.println("Cache hit");
	    }
		
		return result;
		
	}

	public Hashtable<String, String> getLanguagesFromAPI(){
		String response;
		String path = "/languages";
		String query = "?target=" + target  + "&key=" + key;
		String url = baseurl + path + query;
		Hashtable<String,String> hash = new Hashtable<String, String>();
		
		try {
			response = sendGet(url);
			//System.out.println(response);
        } catch (Exception e) {
            //do something clever with the exception
        	response = "{}";
            System.out.println(e.getMessage());
        }	
		
		JsonParser jsonParser = new JsonParser();
		JsonArray arr = jsonParser.parse(response)
						.getAsJsonObject()
						.getAsJsonObject("data")
						.getAsJsonArray("languages");
		
		for (int i = 0; i < arr.size(); i++) {
			JsonElement el = arr.get(i).getAsJsonObject();
			hash.put(	el.getAsJsonObject().get("language").getAsString(), 
						el.getAsJsonObject().get("name").getAsString());
		}
		
		return hash;
	}
	
	private String sendGet(String url) throws Exception {
		 
 
		URL obj = new URL(url);
		HttpURLConnection con = (HttpURLConnection) obj.openConnection();
 
		// optional default is GET
		con.setRequestMethod("GET");
 
 
		int responseCode = con.getResponseCode();
		System.out.println("\nSending 'GET' request to URL : " + url);
		System.out.println("Response Code : " + responseCode);
 
		BufferedReader in = new BufferedReader(
		        new InputStreamReader(con.getInputStream()));
		String inputLine;
		StringBuffer response = new StringBuffer();
 
 		try {
			while ((inputLine = in.readLine()) != null) {
				response.append(inputLine);
			}
			in.close();
 
			//print result
			return response.toString();
 		} catch(Exception ex) {
			ex.printStackTrace();
		} 
		return "";
	}
	
}
